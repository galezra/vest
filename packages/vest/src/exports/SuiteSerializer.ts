import { CB } from 'vest-utils';
import { IsolateSerializer } from 'vestjs-runtime';
import { IsolateKeys } from 'vestjs-runtime/src/Isolate/IsolateKeys';

import { TIsolateSuite } from 'IsolateSuite';
import { TestStatus } from 'IsolateTestStateMachine';
import { TFieldName, TGroupName } from 'SuiteResultTypes';
import { Suite } from 'SuiteTypes';

export type Dumpable = {
  dump: CB<TIsolateSuite>;
};

export class SuiteSerializer {
  static serialize(suite: Dumpable) {
    const dump = { ...suite.dump() };

    return IsolateSerializer.serialize(dump, suiteSerializerReplacer);
  }

  static deserialize(
    serialized: string | TIsolateSuite | Record<string, any>,
  ): TIsolateSuite {
    return IsolateSerializer.deserialize(serialized) as TIsolateSuite;
  }

  static resume(
    suite: Suite<TFieldName, TGroupName>,
    root: string | TIsolateSuite | Record<string, any>,
  ): void {
    const suiteRoot = SuiteSerializer.deserialize(root);

    suite.resume(suiteRoot);
  }
}

function suiteSerializerReplacer(value: any, key: string) {
  if (key === 'output') {
    return undefined;
  }

  if (key === IsolateKeys.Status) {
    if (AllowedStatuses.has(value)) {
      return value;
    }

    return undefined;
  }

  if (DisallowedKeys.has(key)) {
    return undefined;
  }

  return value;
}

const AllowedStatuses = new Set([
  TestStatus.FAILED,
  TestStatus.PASSING,
  TestStatus.WARNING,
]);

const DisallowedKeys = new Set(['focusMode', 'match', 'matchAll', 'severity']);
