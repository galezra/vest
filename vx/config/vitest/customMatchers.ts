import { isFunction } from 'vest-utils';

expect.extend({
  isDeepCopyOf,
});

// eslint-disable-next-line complexity
function isDeepCopyOf(
  source: any,
  clone: any,
): { pass: boolean; message: () => string } {
  const queue = [[source, clone]];

  outer: while (queue.length) {
    const [source, clone] = (queue as any[]).shift();

    if (!source || typeof source !== 'object') {
      if (!isFunction(clone)) {
        if (clone !== source) {
          return {
            pass: false,
            message: () => `Source and clone are not identical`,
          };
        }
      }
      continue;
    }

    if (clone === source) {
      return {
        pass: false,
        message: () =>
          `Source and clone are the same object. Expected a deep copy. ${JSON.stringify(
            source,
          )}===${JSON.stringify(clone)}`,
      };
    }

    // Short circuit
    if (
      (clone && !source) ||
      (source && !clone) ||
      typeof source !== typeof clone
    ) {
      return {
        pass: false,
        message: () =>
          `Source and clone are not of the same type: ${JSON.stringify(
            source,
          )} does not equal ${JSON.stringify(clone)}`,
      };
    }

    if (Array.isArray(source)) {
      // Short circuit
      if (!Array.isArray(clone) || source.length !== clone.length) {
        return {
          pass: false,
          message: () =>
            `source and clone arrays are not identical. ${JSON.stringify(
              source,
            )} does not equal ${JSON.stringify(clone)}`,
        };
      }
      source.forEach((_, i) => {
        queue.push([source[i], clone[i]]);
      });
    } else if (typeof source === 'object') {
      Object.keys(source).forEach(key => queue.push([source[key], clone[key]]));
    }

    continue outer;
  }

  return { pass: true, message: () => 'success' };
}
