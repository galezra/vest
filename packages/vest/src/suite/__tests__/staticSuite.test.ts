import { SuiteSerializer } from 'SuiteSerializer';
import { VestIsolateType } from 'VestIsolateType';
import { describe, it, expect } from 'vitest';
import wait from 'wait';

import * as vest from 'vest';
import { staticSuite } from 'vest';

describe('staticSuite', () => {
  it('Should return a function', () => {
    expect(typeof staticSuite(() => {})).toBe('function');
  });

  it('Should return a "suite instance"', () => {
    const suite = staticSuite(() => {});
    const result = suite();
    expect(typeof result).toBe('object');
    expect(typeof result.tests).toBe('object');
    expect(typeof result.groups).toBe('object');
    expect(typeof result.warnCount).toBe('number');
    expect(typeof result.errorCount).toBe('number');
    expect(typeof result.testCount).toBe('number');
    expect(typeof result.getWarning).toBe('function');
    expect(typeof result.getError).toBe('function');
    expect(typeof result.getErrors).toBe('function');
    expect(typeof result.getErrors).toBe('function');
    expect(typeof result.getWarnings).toBe('function');
    expect(typeof result.getWarnings).toBe('function');
    expect(typeof result.getErrorsByGroup).toBe('function');
    expect(typeof result.getErrorsByGroup).toBe('function');
    expect(typeof result.getWarningsByGroup).toBe('function');
    expect(typeof result.getWarningsByGroup).toBe('function');
    expect(typeof result.hasErrors).toBe('function');
    expect(typeof result.hasWarnings).toBe('function');
    expect(typeof result.hasErrorsByGroup).toBe('function');
    expect(typeof result.hasWarningsByGroup).toBe('function');
    expect(typeof result.isValid).toBe('function');
    expect(typeof result.isValidByGroup).toBe('function');
    expect(typeof result.done).toBe('function');
  });

  it('On consecutive calls, should return a new "suite instance"', () => {
    const suite = staticSuite(only => {
      vest.only(only);
      vest.test('t1', () => false);
      vest.test('t2', () => false);
    });

    const res1 = suite('t1');
    const res2 = suite('t2');

    expect(res1).not.toBe(res2);
    expect(res1.hasErrors('t1')).toBe(true);
    expect(res1.hasErrors('t2')).toBe(false);
    expect(res2.hasErrors('t1')).toBe(false);
    expect(res2.hasErrors('t2')).toBe(true);
  });

  it('Should run async tests normally', () => {
    const suite = staticSuite(() => {
      vest.test('t1', async () => {
        await new Promise(resolve => setTimeout(resolve, 0));
        throw new Error();
      });
    });
    return new Promise<void>(resolve => {
      const res = suite();

      expect(res.hasErrors('t1')).toBe(false);

      res.done(() => {
        resolve();
      });
    });
  });

  describe('dump', () => {
    it('should output a dump of the suite', () => {
      const suite = staticSuite(() => {
        vest.test('t1', () => false);
        vest.test('t2', () => false);

        vest.group('g1', () => {
          vest.test('t1', () => false);
          vest.test('t3', () => false);
        });
      });

      const res = suite();

      expect(res.dump()).toHaveProperty('$type', VestIsolateType.Suite);
      expect(res.dump()).toHaveProperty('children');
      expect(res.dump().children).toHaveLength(3);
      expect(res.dump().children?.[0]).toHaveProperty(
        '$type',
        VestIsolateType.Test,
      );
      expect(res.dump().children?.[1]).toHaveProperty(
        '$type',
        VestIsolateType.Test,
      );
      expect(res.dump().children?.[2]).toHaveProperty(
        '$type',
        VestIsolateType.Group,
      );

      expect(res.dump()).toMatchSnapshot();
    });
  });
});

describe('runStatic', () => {
  it('Should run a static suite', () => {
    const suite = vest.create(() => {
      vest.test('t1', () => false);
      vest.test('t2', () => false);
    });

    const res = suite.runStatic();
    expect(suite.runStatic()).not.toBe(suite.runStatic());

    expect(res.hasErrors('t1')).toBe(true);
    expect(res.hasErrors('t2')).toBe(true);
  });

  it('Should serialize and resume a static suite', () => {
    const suite = vest.create(() => {
      vest.test('t1', () => false);
      vest.test('t2', () => false);
    });

    const serialized = SuiteSerializer.serialize(suite.runStatic());
    const suite2 = vest.create(() => {
      vest.test('t3', () => false);
      vest.test('t4', () => false);
    });

    SuiteSerializer.resume(suite2, serialized);

    expect(suite2.hasErrors('t1')).toBe(true);
    expect(suite2.hasErrors('t2')).toBe(true);
    expect(suite2.hasErrors('t3')).toBe(false);
    expect(suite2.hasErrors('t4')).toBe(false);
  });

  describe('runStatic (promise)', () => {
    it("Should resolve with the suite's result", async () => {
      const suite = vest.create(() => {
        vest.test('t1', async () => {
          await wait(100);
          throw new Error();
        });
      });

      const res = suite.runStatic();
      expect(res.errorCount).toBe(0);
      const result = await res;
      expect(result.errorCount).toBe(1);
    });

    it("Should have a dump method on the resolved suite's result", async () => {
      const suite = vest.create(() => {
        vest.test('t1', async () => {
          await wait(100);
          throw new Error();
        });
      });

      const res = suite.runStatic();
      const result = await res;
      expect(result).toHaveProperty('dump');
      expect(result.dump()).toHaveProperty('$type', VestIsolateType.Suite);
    });
  });
});
