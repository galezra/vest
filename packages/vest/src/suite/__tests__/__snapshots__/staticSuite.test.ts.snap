// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`staticSuite > dump > should output a dump of the suite 1`] = `
{
  "$type": "Suite",
  "abortController": AbortController {},
  "allowReorder": undefined,
  "children": [
    {
      "$type": "Test",
      "abortController": AbortController {},
      "allowReorder": undefined,
      "children": null,
      "data": {
        "fieldName": "t1",
        "severity": "error",
        "testFn": [Function],
      },
      "key": null,
      "keys": null,
      "output": undefined,
      "parent": [Circular],
      "status": "FAILED",
    },
    {
      "$type": "Test",
      "abortController": AbortController {},
      "allowReorder": undefined,
      "children": null,
      "data": {
        "fieldName": "t2",
        "severity": "error",
        "testFn": [Function],
      },
      "key": null,
      "keys": null,
      "output": undefined,
      "parent": [Circular],
      "status": "FAILED",
    },
    {
      "$type": "Group",
      "abortController": AbortController {},
      "allowReorder": undefined,
      "children": [
        {
          "$type": "Test",
          "abortController": AbortController {},
          "allowReorder": undefined,
          "children": null,
          "data": {
            "fieldName": "t1",
            "groupName": "g1",
            "severity": "error",
            "testFn": [Function],
          },
          "key": null,
          "keys": null,
          "output": undefined,
          "parent": [Circular],
          "status": "SKIPPED",
        },
        {
          "$type": "Test",
          "abortController": AbortController {},
          "allowReorder": undefined,
          "children": null,
          "data": {
            "fieldName": "t3",
            "groupName": "g1",
            "severity": "error",
            "testFn": [Function],
          },
          "key": null,
          "keys": null,
          "output": undefined,
          "parent": [Circular],
          "status": "FAILED",
        },
      ],
      "data": {},
      "key": null,
      "keys": null,
      "output": undefined,
      "parent": [Circular],
      "status": "DONE",
    },
  ],
  "data": {
    "optional": {},
  },
  "key": null,
  "keys": null,
  "output": {
    "done": [Function],
    "errorCount": 3,
    "errors": [
      SummaryFailure {
        "fieldName": "t1",
        "groupName": undefined,
        "message": undefined,
      },
      SummaryFailure {
        "fieldName": "t2",
        "groupName": undefined,
        "message": undefined,
      },
      SummaryFailure {
        "fieldName": "t3",
        "groupName": "g1",
        "message": undefined,
      },
    ],
    "getError": [Function],
    "getErrors": [Function],
    "getErrorsByGroup": [Function],
    "getMessage": [Function],
    "getWarning": [Function],
    "getWarnings": [Function],
    "getWarningsByGroup": [Function],
    "groups": {
      "g1": {
        "t1": SummaryBase {
          "errorCount": 0,
          "errors": [],
          "pendingCount": 0,
          "testCount": 0,
          "valid": false,
          "warnCount": 0,
          "warnings": [],
        },
        "t3": SummaryBase {
          "errorCount": 1,
          "errors": [],
          "pendingCount": 0,
          "testCount": 1,
          "valid": false,
          "warnCount": 0,
          "warnings": [],
        },
      },
    },
    "hasErrors": [Function],
    "hasErrorsByGroup": [Function],
    "hasWarnings": [Function],
    "hasWarningsByGroup": [Function],
    "isPending": [Function],
    "isTested": [Function],
    "isValid": [Function],
    "isValidByGroup": [Function],
    "pendingCount": 0,
    "suiteName": undefined,
    "testCount": 3,
    "tests": {
      "t1": {
        "errorCount": 1,
        "errors": [],
        "pendingCount": 0,
        "testCount": 1,
        "valid": false,
        "warnCount": 0,
        "warnings": [],
      },
      "t2": SummaryBase {
        "errorCount": 1,
        "errors": [],
        "pendingCount": 0,
        "testCount": 1,
        "valid": false,
        "warnCount": 0,
        "warnings": [],
      },
      "t3": SummaryBase {
        "errorCount": 1,
        "errors": [],
        "pendingCount": 0,
        "testCount": 1,
        "valid": false,
        "warnCount": 0,
        "warnings": [],
      },
    },
    "valid": false,
    "warnCount": 0,
    "warnings": [],
  },
  "parent": null,
  "status": "DONE",
}
`;
